amethyst_animation/src/resources.rs

```diff
@@ -0,0 +1,193 @@
+use std::time::{Duration, Instant};
+
+use amethyst_assets::{Asset, Handle, Result};
+use fnv::FnvHashMap;
+use specs::{Component, DenseVecStorage, Entity};
+
+use interpolation::InterpolationType;
+
+/// The actual animation data for a single attribute
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum AnimationOutput {
+    /// Translation is a 3d vector
+    Translation(Vec<[f32; 3]>),
+    /// Rotation is a quaternion
+    Rotation(Vec<[f32; 4]>),
+    /// Scale is a 3d vector
+    Scale(Vec<[f32; 3]>),
+}
+
+/// Sampler defines a single animation for a single attribute of the `LocalTransform` of the entity
+/// it is attached to.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Sampler {
+    /// Time of key frames
+    pub input: Vec<f32>,
+    /// Actual output data to interpolate
+    pub output: AnimationOutput,
+    /// How should interpolation be done
+    pub ty: InterpolationType,
+}
+
+impl Asset for Sampler {
+    type Data = Self;
+    type HandleStorage = DenseVecStorage<Handle<Self>>;
+}
+
+impl Into<Result<Sampler>> for Sampler {
+    fn into(self) -> Result<Sampler> {
+        Ok(self)
+    }
+}
+
+/// Defines the hierarchy of nodes that a single animation can control.
+/// Attach to the root entity that an animation can be defined for.
+/// Only required for animations which targets more than a single node.
+#[derive(Debug, Clone)]
+pub struct AnimationHierarchy {
+    pub nodes: FnvHashMap<usize, Entity>,
+}
+
+impl Component for AnimationHierarchy {
+    type Storage = DenseVecStorage<Self>;
+}
+
+/// Defines a single animation.
+/// Defines relationships between the node index in `AnimationHierarchy` and a `Sampler` handle.
+/// If the animation only targets a single node index, `AnimationHierarchy` is not required.
+#[derive(Clone, Debug)]
+pub struct Animation {
+    /// node index -> sampler handle
+    pub nodes: Vec<(usize, Handle<Sampler>)>,
+}
+
+impl Asset for Animation {
+    type Data = Self;
+    type HandleStorage = DenseVecStorage<Handle<Self>>;
+}
+
+impl Into<Result<Animation>> for Animation {
```

Comment:
For asset processor.

Reply:
You've answered only first question.

Reply:
I'm using the `Processor` from the assets crate to do asset processing for Sampler and Animation, and it has Asset::Data = Animation (it converts into itself since it doesn't need a factory to convert or such), and that requires an Into<assets::Result<Asset>> for Asset::Data

Reply:
Yes, I could do manual processing of them in my own system, just felt silly to build a new system for something that already exists.